File: Driver.cpp
/* Honor Pledge: I pledge that I have neither 
 given nor receieved any help on this assignment.
Assignment 6: Polymorphism
 C++ Header file
 David Haines */
 
 //imports/includes for classes
 #include "Driver.h"
 #include "Sort.h"
 #include "Bubble.h"
 #include "Insertion.h"
 
 //imports/includes for operations
 #include <fstream>
 #include <sstream>
 #include <string>
 #include <stdlib.h>
 #include <iostream>
 
 
 ////----Driver Main Function----////
 void Driver::run()
 {
	//exit loop variable
	int menuOne = 0;
	
	///---First Menu---///
	while (menuOne == 0)
	{
		//input variable declaration
		std::string input = "";
		
		//print menu and ask user to enter which one they would like to do
		std::cout << "" << std::endl;
		std::cout << "1) Load Data (From File)" << std::endl;
		std::cout << "2) Exit Program" << std::endl;
		std::cout << "Please Enter Your Selection: " << std::endl;
		std::cin >> input;
		
		//if, else if statement for determining which option user wanted;
		//if user entered incorrect entry, tell them to re-enter their selection
		if (input == "1")
		{
			//run loadFile function to load in the text file and convert string to numbers
			loadFile();
			
			//exit loop variable for second menu
			int menuTwo = 0;
			
			///---Second Menu---///
			while (menuTwo == 0)
			{
				//second menu input variable declaration
				std::string inputTwo = "";
				
				//print menu and ask user to enter which one they would like to do
				std::cout << "" << std::endl;
				std::cout << "1) Insertion Sort" << std::endl;
				std::cout << "2) Bubble Sort" << std::endl;
				std::cout << "3) Exit Program" << std::endl;
				std::cout << "Please Enter Your Selection: " << std::endl;
				std::cin >> inputTwo;
				
				//determine which option user selected, or if entry was invalid; do said option
				if (inputTwo == "1")
				{
					//create Insertion object
					Sort * insSort = new Insertion();
					
					//call the sort Array method//
					insSort->sortArray(numbers);
					
					//print sorted array
					std::cout << "" << std::endl;
					std::cout << "Insertion Sort: " << arrayConcat() << std::endl;
					
					//delete Insertion Sort
					delete insSort;
					
					//back to main menuOne
					menuTwo = 1;
				}
				else if (inputTwo == "2")
				{
					//create Bubble object
					Sort * bubSort = new Bubble();
					
					//call the sort Array method//
					bubSort->sortArray(numbers);
					
					//print sorted array
					std::cout << "" << std::endl;
					std::cout << "Bubble Sort: " << arrayConcat() << std::endl;
					
					//delete Bubble sort object
					delete bubSort;
					
					//back to first menu
					menuTwo = 1;
				}
				else if (inputTwo == "3")
				{
					//exit both menus/program
					menuOne = 1;
					menuTwo = 1;
				}
				else
				{
					//invalid input, give message and go back to second menu start
					std::cout << "" << std::endl;
					std::cout << "I'm not sure what you were trying to do." << std::endl;
					std::cout << "Please try entering 1, 2, or 3."  << std::endl;
				}
			}
		}
		else if (input == "2")
		{
			//exit while loop/program
			menuOne = 1;
		}
		else 
		{
			std::cout << "" << std::endl;
			std::cout << "I'm not sure what you were trying to do." << std::endl;
			std::cout << "Please try entering 1 or 2."  << std::endl;
		}
		
		//exit message
		std::cout << "Ok, exiting now. Goodbye!" << std::endl;
	}
	
	return;
 }
 
 
 ///---Load File Function---///
 void Driver::loadFile()
 {	
	//Extracted line variable
	std::string line;
	//create int that will be the converted input string
	int number;
	//create string that holds individual number in csv file
	std::string strInt = "";
	
	//Stringstream instance for reading in and delimiting
	std::stringstream ss;
	//Stringstream instance for converting
	std::stringstream converter;
	
	//Input Stream Variable (opens file, too)
	std::ifstream inputFile("data.txt");
	
	
	//see if it is open; if so, read in numbers and place in array
	if (inputFile.is_open())
	{		
		
		//--while loop to get each line--//
		while (std::getline(inputFile, line))
		{
			//clear ss object and ready it for use
			ss.clear();
			ss.str("");
			
			//set ss object to take in current line
			ss.str(line);
			
			//declare numbers array index variable
			int i = 0;
			
			//--second while loop to parse through comma delimited values--//
			/*
			*Takes in ss value (read-in line), sepearates field based on commas in line,
			*loops through all fields, sets strInt to field value, convert and put in array,
			*then moves to next field
			*/
			while (std::getline(ss, strInt, ','))
			{
				//clear converter from last field value
				converter.clear();
				converter.str("");
			
				//convert string to decimal/int: take in the string variable for each csv field, put out into int variable
				converter << strInt;
				converter >> number;
				
				//set the output number to array element with index i and increment i
				numbers[i] = number;
				i++;
			}
		}
		
		//close the input stream
		inputFile.close();
		
		//print array with array concatenation function	
		std::cout << "" << std::endl;
		std::cout << "File imported." << std::endl;
		std::cout << "Unsorted Array: " << arrayConcat() << std::endl;
		
	}
	else
	{
		//if file can't be opened, print message to user.
		std::cout << "Unable to open file" << std::endl;
	}
	
	return;
 }

 
 ////----Array String Concatenation For Printing----////
 std::string Driver::arrayConcat()
 {
	//variable for index
	int index = 0;
	
	//variable for array string concat
	std::string concatArray = "";
	
	//create stringstream to input number at array index then convert to str
	std::stringstream arrayNum;
	
	//convert int at numbers[i] to a string, add it to the string of all array numbers with a comma and space after
	for (index; index < 50; index++)
	{
		//clear stringstream
		arrayNum.clear();
		arrayNum.str("");
		
		//take in the number at index
		arrayNum << numbers[index];
		
		//set string to arrayNum 
		std::string s = arrayNum.str();
		
		/*
		*std::string s = std::to_string (numbers[index]);
		*DOESN'T WORK WITH THIS COMPILER
		*/
		
		//add the array element to the string with a comma and space divider
		concatArray = concatArray + s + ", ";
	}

	//erase last comma and space in strArray
	std::string strArray = concatArray.substr(0, strArray.length() - 2);

	//return string with all array elements concatenated, without last space and comma
	return strArray;
 }
 
 ////----MAIN Program----////
 int main()
 {
	//create new instance of driver and call the main (run) function
	Driver * drive = new Driver();
	drive->run();
	
	//delete drive when done and return 0
	delete drive;
	return 0;
 }

File: Java version/Driver.java
/* 
*Honor Pledge: I pledge that I have neither 
*given nor receieved any help on this assignment.
*Assignment 6: Polymorphism
*Driver Class
*David Haines 
*/

///---Imports---///
//import java io for file input readers, exceptions, and possible formatted output (might just use system.out)
import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.PrintStream;
import java.util.Scanner;

///---Driver Class---///
public class Driver
{
	//declare numbers array as private static
	public static int[] numbers = new int[50];
	
	
	///---Main---///
	public static void main(String[] args)
	{
		//exit loop variable
		int menuOne = 0;
	
		///---First Menu---///
		while (menuOne == 0)
		{
			//Set up scanner input for first menu
			Scanner input = new Scanner(System.in);
		
			//print menu and ask user to enter which one they would like to do
			System.out.println("");
			System.out.println("1) Load Data (From File)");
			System.out.println("2) Exit Program");
			System.out.printf("%nPlease Enter Your Selection: ");
			String in = input.next();
		
			//if, else if statement for determining which option user wanted;
			//if user entered incorrect entry, tell them to re-enter their selection
			if (in.equals("1"))
			{
				//run loadFile function to load in the text file and convert string to numbers
				loadFile();
			
				//exit loop variable for second menu
				int menuTwo = 0;
			
				///---Second Menu---///
				while (menuTwo == 0)
				{
					
					//Set up scanner input for first menu
					Scanner inputTwo = new Scanner(System.in);
				
					//print menu and ask user to enter which one they would like to do
					System.out.println("");
					System.out.println("1) Insertion Sort");
					System.out.println("2) Bubble Sort");
					System.out.println("3) Exit Program");
					System.out.printf("%nPlease Enter Your Selection: ");
					String inp = inputTwo.next();
				
					//determine which option user selected, or if entry was invalid; do said option
					if (inputTwo.equals("1"))
					{
						//create Insertion object
						Sort insSort = new Insertion();
					
						//call the sort Array method//
						insSort.sortArray(numbers);
					
						//print sorted array
						System.out.printf("%nInsertion Sort: %s%n", arrayConcat());
					
						//back to main menuOne
						menuTwo = 1;
					}
					else if (inputTwo.equals("2"))
					{
						//create Bubble object
						Sort bubSort = new Bubble();
					
						//call the sort Array method//
						bubSort.sortArray(numbers);
					
						//print sorted array
						System.out.printf("%nBubble Sort: %s%n", arrayConcat());
					
						//back to first menu
						menuTwo = 1;
					}
					else if (inputTwo.equals("3"))
					{
						//exit both menus/program
						menuOne = 1;
						menuTwo = 1;
					}
					else
					{
						//invalid input, give message and go back to second menu start
						System.out.println("");
						System.out.println("I'm not sure what you were trying to do.");
						System.out.println("Please try entering 1, 2, or 3.");
					}
				}
			}
			else if (in.equals("2"))
			{
				//exit while loop/program
				menuOne = 1;
			}
			else 
			{
				System.out.println("");
				System.out.println("I'm not sure what you were trying to do.");
				System.out.println("Please try entering 1 or 2.");
			}
		}
	
		//exit message
		System.out.println("");
		System.out.println("Ok, exiting now. Goodbye!");
		System.out.println("");
	}
 
 
	///---Load File Function---///
	public static void loadFile()
	{	
		try
		{
			//input stream for text file, input stream reader then buffer reader
			InputStream inpStrm = new FileInputStream("data.txt");
			
			InputStreamReader isr = new InputStreamReader(inpStrm);
			
			BufferedReader br = new BufferedReader(isr);
			
			//burn first line of file (header)
			String line = br.readLine();		
		
			//--while loop to get each line--//
			while (line != null)
			{
				//put line split into numbers array
				String numbs[] = line.split(",");
			
				//--second while loop to convert string array to int array--//
			
				for (int i = 0; i < numbs.length; i++)
				{
					//parse the string in index i spot for an int, put that value in int numbers array with same index
					numbers[i] = Integer.parseInt(numbs[i]);
				}
				
				//get next line
				line = br.readLine();
			}
		
			//close the input stream
			br.close();
		
			//print array with array concatenation function	
			System.out.println("");
			System.out.println("File imported.");
			System.out.printf("%nUnsorted Array: %s%n", arrayConcat());
		
		}
		catch (IOException ex)
		{
			//if file can't be opened, print message to user.
			System.out.println("Unable to open file");
			System.err.println(ex);
		}
	
	}

 
	////----Array String Concatenation For Printing----////
	public static String arrayConcat()
	{
	
		//variable for array string concat
		String concatArray = "";
	
		//convert int at numbers[i] to a string, add it to the string of all array numbers with a comma and space after
		for (int index = 0; index < 50; index++)
		{
			int i = numbers[index];
		
			//add the array element to the string with a comma and space divider
			concatArray = concatArray + String.valueOf(i) + ", ";
		}

		//erase last comma and space in strArray
		String strArray = concatArray.substring(0, concatArray.length() - 2);

		//return string with all array elements concatenated, without last space and comma
		return strArray;
	}
}